payment/src/main/java/org/killbill/billing/payment/core/sm/control/ControlPluginRunner.java

```diff
@@ -122,12 +126,12 @@ public PriorPaymentControlResult executePluginPriorCalls(final Account account,
             }
 
             if (prevResult.getAdjustedPaymentMethodId() != null) {
-                // We only allow setting the paymentMethodId but disallow overwriting an existing paymentMethodId for a given Payment. See #1097
-                if (paymentMethodId != null) {
+                // We only allow setting the paymentMethodId but disallow overwriting an existing paymentMethodId for a given Payment - See #1097
+                // unless the property isAllowedToOverwritePaymentMethodId was explicitly configured to allow this
+                if (paymentConfig.isAllowedToOverwritePaymentMethodId() || paymentMethodId != null) {
```

Comment:
Shouldn't it be?

```
!paymentConfig.isAllowedToOverwritePaymentMethodId() && paymentMethodId != null
```

Reply:
If the plugin is trying to set/update the payment method - line 128- then:
1. Either we allow to update/set the payment method - first part of the `if` with `paymentConfig.isAllowedToOverwritePaymentMethodId()`
2. OR (the config flag to update update/set the payment method is NOT on), but we check that it is *SET* and not an *UPDATE* and then we proceed - second part of the if `paymentMethodId != null`

Reply:
But the `if` clause guards the `throw`?

Reply:
🤦‍♂️ Fixed in https://github.com/killbill/killbill/pull/1559/commits/015a7752366dfe0a4f76f274e77d3770965b45ea